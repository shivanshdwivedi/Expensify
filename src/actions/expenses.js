// jshint esversion : 6
// jshint esversion : 9

import uuid from 'uuid';
import database from '../Firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch , getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({id} = {}) => {
  return(dispatch , getState) => {

    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id , updates) => {
  return(dispatch , getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id , updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// export const startSetexpense;

// 1. Fetch all expense data once
// 2. Parse that data into an array
// 3. Dispatch SET_EXPENSES

export const startSetExpenses = () => {
  return (dispatch , getState) => {

    const uid = getState().auth.uid

    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];

      snapshot.forEach((childsnapshot) => {
        expenses.push({
          id: childsnapshot.key,
          ...childsnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};

// startRemoveExpense