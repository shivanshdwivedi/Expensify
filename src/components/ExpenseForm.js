// jshint esversion : 6
// jshint esversion : 9


import React from 'react';

export default class ExpenseForm extends React.Component {
      state = {
        description: '',
        note: ''
      };
      onDescriptionChange = (e) => {
      const description = e.target.value;
      this.setState(() => ({ description }));
      };

      onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
      }

    render() {
        return (
            <div>
                Expense Form
                <form>
                    <input
                    type = 'text'
                    placeholder = 'Description'
                    autofocus
                    value = {this.state.description}
                    onChange = {this.onDescriptionChange}
                    />
                    <input
                    type = 'number'
                    placeholder = 'Amount' 
                    />
                    <div>
                    <textarea
                    placeholder = 'Add a note for your expense(optional)'
                    value = {this.state.note}
                    onChange = {this.onNoteChange}
                    />
                    </div>
                    
                    <div>
                    <button>Add Expense</button>
                    </div>
                    

                </form>
            </div>
        )
    }
}