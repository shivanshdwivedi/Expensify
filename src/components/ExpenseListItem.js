// Export a stateless functional component
// description , amout , createdAt
import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

const ExpenseListItem = ({description , amount , createdAt ,  id }) => (
    <div>
    <Link to ={`/edit/${id}`}>   
    <h3> {description} </h3>
    </Link>
        <p>
        {numeral(amount/100).format('Rs0,0.00')} 
        - 
        {moment(createdAt).format('MMMM Do , YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;