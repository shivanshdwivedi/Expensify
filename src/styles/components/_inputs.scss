.text-input , .select{
    border: 1px solid #cacccd;
    height: 50;
    font-size: $font-size-large;
    font-weight: 300;
    padding: $s-size;
}

.select {
    @extend .text-input;;
}

.textarea {
    @extend .text-input;
    height: 10rem;
}