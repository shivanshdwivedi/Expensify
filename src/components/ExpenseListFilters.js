// jshint esversion : 6
// jshint esversion : 9

import React from 'react';
import {connect} from 'react-redux';
import { setTextFilter, setEndDate, setStartDate } from '../actions/filters';
import { sortByDate } from '../actions/filters';
import { sortByAmount , startDate ,endDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

class ExpenseListFilters extends React.Component{
    state = {
        calenderFocused: null
    };
    onDatesChange = ({startDate , endDate}) => {
     this.props.dispatch(setStartDate(startDate));
     this.props.dispatch(setEndDate(endDate));
    };
    onFocusChange = (calenderFocused) => {
    this.setState(() => ({calenderFocused}));
    }
    render(){
        return(
            <div>
                <input type = 'text' value = {this.props.filters.text} onChange = {(e) => {
                    props.dispatch(setTextFilter(e.target.value))
                }}/>
                <select 
                    value = {this.props.filters.sortBy}  
                    onChange = {(e) =>{
                    if(e.target.value === 'date'){
                        props.dispatch(sortByDate());
                    }else if(e.target.value === 'amount'){
                     props.dispatch(sortByAmount());
                    }
                     }}
                >
                     <option value = 'amount'>Amount</option>
                    <option value = 'date'>Date</option>
                </select>
              <DateRangePicker 
                  startDate = {this.props.filters.startDate}
                  endDate = {this.props.filters.endDate}
                  onDatesChange= {this.onDatesChange}
                  focusedInput = {this.state.calenderFocused}
                  onFocusChange = {this.onFocusChange}
                  showClearDates = {true}
                  numberOfMonths = {1}
                  isOutsideRange = {() => false}
              />

            </div>
        )

    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters);