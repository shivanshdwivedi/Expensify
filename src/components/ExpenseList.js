import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
const ExpenseList = (props) => (
    <div>
        <h1>
            Expense List
        </h1>
            {ExpenseListItem}
    </div>
);

const mapStateToProps = (state) => {
  return{
   expenses: state.expenses,
   filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseList)

