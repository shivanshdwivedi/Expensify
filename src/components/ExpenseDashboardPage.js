// jshint esversion : 6
// jshint esversion : 9

import React from 'react';
import ExpenseList  from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
       <ExpenseListFilters />
       <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;