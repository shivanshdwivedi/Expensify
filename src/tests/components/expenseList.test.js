import {ExpenseList} from '../../components/ExpenseList';
import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';

test('' , () => {
    const wrapper = shallow(<ExpenseList expenses = {expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('' , () => {
    const wrapper = shallow(<ExpenseList expenses = {[]} />);
    expect(wrapper).toMatchSnapshot();
});