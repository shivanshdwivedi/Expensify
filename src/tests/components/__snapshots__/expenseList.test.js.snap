// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="gum"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={19500}
    createdAt={-345600000}
    description="Pencil"
    id={2}
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={100}
    createdAt={345600000}
    description="Earser"
    id={3}
    key="3"
    note=""
  />
</div>
`;

exports[` 2`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
